File: core/src/main/java/io/github/bindglam/neko/registry/ScalableRegistry.java
package io.github.bindglam.neko.registry;

import it.unimi.dsi.fastutil.Pair;
import net.kyori.adventure.key.Key;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Unmodifiable;
import java.util.*;

public abstract class ScalableRegistry<T> implements WritableRegistry<T> {
    protected final Map<String, T> map = new HashMap<>();
    protected volatile boolean isLocked = false;

    @Override
    public synchronized void lock() {
        isLocked = true;
    }

    @Override
    public synchronized void unlock() {
        isLocked = false;
    }

    @Override
    public void merge(@NotNull Registry<T> registry) {
        if (isLocked) {
            throw new IllegalStateException("The registry is locked");
        }

        Map<String, T> newMap = new HashMap<>(map);
        for (Pair<Key, T> entry : registry.entries()) {
            String key = entry.key().asString();
            if (newMap.containsKey(key)) {
                throw new IllegalStateException("There is conflict with the registry");
            }
            newMap.put(key, entry.value());
        }
        map.clear();
        map.putAll(newMap);
    }

    @Override
    public void clear() {
        if (isLocked) {
            throw new IllegalStateException("The registry is locked");
        }
        map.clear();
    }

    @Override
    public @NotNull Optional<T> get(@NotNull Key key) {
        return Optional.ofNullable(map.get(key.asString()));
    }

    @Override
    public @NotNull @Unmodifiable Collection<Pair<Key, T>> entries() {
        List<Pair<Key, T>> result = new ArrayList<>();
        for (Map.Entry<String, T> entry : map.entrySet()) {
            result.add(Pair.of(Key.key(entry.getKey()), entry.getValue()));
        }
        return Collections.unmodifiableList(result);
    }

    @Override
    public @NotNull Iterator<T> iterator() {
        return map.values().iterator();
    }
}

File: core/src/main/kotlin/io/github/bindglam/neko/registry/ScalableRegistry.kt
package io.github.bindglam.neko.registry

import it.unimi.dsi.fastutil.Pair
import net.kyori.adventure.key.Key
import java.util.Optional
import java.util.function.Consumer

class ScalableRegistry<T, E : WritableRegistry.RegistryEntry<T>>(private val entrySupplier: () -> E) : WritableRegistry<T, E> {
    private val map = hashMapOf<String, T>()

    private var isLocked = false

    @Synchronized
    override fun lock() {
        isLocked = true
    }

    @Synchronized
    override fun unlock() {
        isLocked = false
    }

    override fun register(key: Key, entry: Consumer<E>): T & Any {
        if(isLocked)
            error("The registry is locked")
        if(map.contains(key.asString()))
            error("The registry already contains the key")
        return entrySupplier().apply { entry.accept(this) }.toValue().also { map[key.asString()] = it }
    }

    override fun register(key: Key, value: T & Any): T & Any {
        if(isLocked)
            error("The registry is locked")
        if(map.contains(key.asString()))
            error("The registry already contains the key")
        return value.also { map[key.asString()] = it }
    }

    override fun merge(registry: Registry<T>) {
        if(isLocked)
            error("The registry is locked")

        val newMap = HashMap(map)
        for (entry in registry.entries()) {
            if(newMap.containsKey(entry.key().asString()))
                error("There is conflict with the registry")
            newMap[entry.key().asString()] = entry.value()
        }
        map.clear()
        map.putAll(newMap)
    }

    override fun get(key: Key) = Optional.ofNullable(map[key.asString()])

    override fun entries() = map.entries.map { entry -> Pair.of(Key.key(entry.key), entry.value) }

    override fun iterator() = map.values.iterator()
}